File: LinkedList/java/leetcode138_Copy_List_with_Random_Pointer.java
//A linked list is given such that each node contains an additional random pointer which could point to any node in the list or null.
//
//Return a deep copy of the list.

package wenxinjie2;

import java.util.HashMap;

public class leetcode138_Copy_List_with_Random_Pointer {
    public RandomListNode copyRandomList(RandomListNode head) {
        if (head == null) return null;
        HashMap<RandomListNode, RandomListNode> map = new HashMap<RandomListNode, RandomListNode>();
        RandomListNode clonehead = new RandomListNode(head.label);
        RandomListNode clonecur = clonehead, cur = head;
        while (cur != null){
            if (cur.next != null){
                if (map.containsKey(cur.next)) clonecur.next = map.get(cur.next);
                else{
                    RandomListNode tem = new RandomListNode(cur.next.label);
                    clonecur.next = tem;
                    map.put(cur.next, clonecur.next);
                }
            }
            if (cur.random != null){
                if (map.containsKey(cur.random)) clonecur.random = map.get(cur.random);
                else{
                    RandomListNode tem1 = new RandomListNode(cur.random.label);
                    clonecur.random = tem1;
                    map.put(cur.random, clonecur.random);
                }
            }
            cur = cur.next;
            clonecur = clonecur.next;
        }
        return clonehead;
    }
    
    
    public RandomListNode copyRandomList1(RandomListNode head) {
        if(head == null) return null;
        RandomListNode h = head;
        while (h != null){
            RandomListNode cur = new RandomListNode(h.label);
            cur.next = h.next;
            h.next = cur;
            h = cur.next;
        }
        h = head;
        while (h != null){
            if (h.random!= null) h.next.random = h.random.next;
            h = h.next.next;
        }
        h = head;
        RandomListNode head1 = new RandomListNode(0);
        RandomListNode newhead = head1;
        while (h != null){
            newhead.next = h.next;
            newhead = newhead.next;
            RandomListNode t = h.next.next;
            h.next = t;
            h = h.next;              
        }
        return head1.next;
    }   
}

//Time: O(n)
//Space: O(n)
//Difficulty: medium

File: LinkedList/python/leetcode138_Copy_List_with_Random_Pointer.py
# A linked list is given such that each node contains an additional random pointer which could point to any node in the list or null.

# Return a deep copy of the list.


# Definition for singly-linked list with a random pointer.
# class RandomListNode(object):
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        """
        if not head:  return None
        clonehead = RandomListNode(head.label)
        clonecur = clonehead
        cur = head
        dic = {head:clonehead}
        while cur:
            if cur.next:
                if cur.next in dic:
                    clonecur.next = dic[cur.next]
                else:
                    clonecur.next = RandomListNode(cur.next.label)
                    dic[cur.next] = clonecur.next
            
            if cur.random:
                if cur.random in dic:
                    clonecur.random = dic[cur.random]
                else:
                    clonecur.random = RandomListNode(cur.random.label)
                    dic[cur.random] = clonecur.random
            cur = cur.next
            clonecur = clonecur.next
        return clonehead


    def copyRandomList2(self, head):
    	if not head: return head
        h = head
        while h:
            n = RandomListNode(h.label)
            n.next = h.next
            h.next = n
            h = n.next
        h = head
        while h:
            if h.random:
                h.next.random = h.random.next
            h = h.next.next
        nh = dummy = RandomListNode(0)
        h = head
        while h:
            nh.next = h.next
            nh = nh.next            
            t = h.next.next
            h.next = t
            h = h.next

        return dummy.next

# Time: O(n)
# Space: O(n)
# Difficulty: medium